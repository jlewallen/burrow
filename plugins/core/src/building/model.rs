use crate::library::model::*;

use std::str::FromStr;

#[derive(Clone, Serialize, Deserialize, PartialEq, ToTaggedJson, Reply, Debug)]
#[serde(rename_all = "camelCase")]
pub enum Constructed {
    Area(AreaObservation),
}

impl TryFrom<Constructed> for Effect {
    type Error = TaggedJsonError;

    fn try_from(value: Constructed) -> std::result::Result<Self, Self::Error> {
        Ok(Self::Reply(value.to_tagged_json()?.into()))
    }
}

#[derive(Debug, Serialize, ToTaggedJson)]
#[serde(rename_all = "camelCase")]
struct EditorReply {}

impl Reply for EditorReply {}

#[derive(Default, Clone, Debug)]
pub struct QuickEdit {
    pub name: Option<String>,
    pub desc: Option<String>,
}

impl TryFrom<&EntityPtr> for QuickEdit {
    type Error = DomainError;

    fn try_from(value: &EntityPtr) -> std::result::Result<Self, Self::Error> {
        let name = Some(value.name()?);
        let desc = value.desc()?;
        Ok(Self { name, desc })
    }
}

const SEPARATOR: &str =
    "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~";

impl FromStr for QuickEdit {
    type Err = DomainError;

    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
        let parts: Vec<_> = s.split(SEPARATOR).collect();
        if parts.len() != 2 {
            return Err(DomainError::Anyhow(anyhow::anyhow!("malformed quick edit")));
        }

        let (name, desc) = match parts[..] {
            [name, desc] => (name, desc),
            _ => todo!(),
        };

        let name = Some(name.trim().to_owned());
        let desc = Some(desc.trim().to_owned());

        Ok(Self { name, desc })
    }
}

impl ToString for QuickEdit {
    fn to_string(&self) -> String {
        format!(
            "{}\n\n{}\n\n{}",
            self.name.as_deref().unwrap_or(""),
            SEPARATOR,
            self.desc.as_deref().unwrap_or("")
        )
    }
}
