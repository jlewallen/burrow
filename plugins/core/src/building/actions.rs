use std::{rc::Rc, str::FromStr};

use chrono::Utc;

use crate::{
    building::model::{Constructed, QuickEdit},
    carrying::model::{Carryable, Containing},
    library::actions::*,
    looking::{actions::LookAction, model::new_area_observation},
    memory::model::{remember, EntityEvent, MemoryEvent},
    moving::model::Occupyable,
};

#[action]
pub struct AddScopeAction {
    pub scope_key: String,
}

impl Action for AddScopeAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, _session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        // TODO Security wise we would probably need a global list of who could
        // manipulate which scopes eventually.
        // TODO Right now this requires scopes to be functionable if all their
        // fields are ommitted. Look into `#[serde(default)]` to make this work
        // w/o a bunch of Option's?
        let Some(item) = tools::holding_one_item(surroundings.living())? else {
            return Ok(SimpleReply::NotFound.try_into()?);
        };

        debug!(item = ?item, scope_key = %self.scope_key, "add-scope");

        let mut item = item.entity().borrow_mut();

        item.add_scope_by_key(&self.scope_key);

        Ok(SimpleReply::Done.try_into()?)
    }
}

#[action]
pub struct EditAction {
    pub item: Item,
}

impl Action for EditAction {
    fn is_read_only() -> bool {
        true
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        debug!("editing {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(editing) => {
                info!("editing {:?}", editing);
                let quick_edit: QuickEdit = (&editing).try_into()?;
                Ok(EditorReply::new(
                    editing.key().to_string(),
                    WorkingCopy::Markdown(quick_edit.to_string()),
                    SaveQuickEditAction::new_template(editing.key().clone())?,
                )
                .try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct EditRawAction {
    pub item: Item,
}

impl Action for EditRawAction {
    fn is_read_only() -> bool {
        true
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        debug!("editing {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(editing) => {
                info!("editing {:?}", editing);
                let json = {
                    let editing = editing.borrow();
                    editing.to_json_value()?
                };
                let key = editing.key().clone();
                Ok(EditorReply::new(
                    key.to_string(),
                    WorkingCopy::Json(json),
                    SaveEntityJsonAction::new_template(key)?,
                )
                .try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct DuplicateAction {
    pub item: Item,
}

impl Action for DuplicateAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("duplicating {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(duplicating) => {
                info!("duplicating {:?}", duplicating);
                _ = tools::duplicate(&duplicating)?;
                Ok(SimpleReply::Done.try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct ObliterateAction {
    pub item: Item,
}

impl Action for ObliterateAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("obliterate {:?}!", self.item);

        match session.find_item(surroundings, &self.item)? {
            Some(obliterating) => {
                info!("obliterate {:?}", obliterating);
                tools::obliterate(&obliterating)?;
                Ok(SimpleReply::Done.try_into()?)
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct MakeItemAction {
    pub name: String,
}

impl Action for MakeItemAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("make-item {:?}", self.name);

        let creator = surroundings.living();

        let new_item: Entity = build_entity()
            .default_scope::<Carryable>()?
            .creator(creator.entity_ref())
            .name(&self.name)
            .try_into()?;

        let new_item = session.add_entity(new_item)?;

        tools::set_quantity(&new_item, 1f32)?;
        tools::set_container(creator, &vec![new_item.clone()])?;

        remember(
            &creator,
            Utc::now(),
            MemoryEvent::Created(EntityEvent {
                key: new_item.key().clone(),
                gid: new_item.gid(),
                name: new_item.name()?.unwrap(),
            }),
        )?;

        Ok(SimpleReply::Done.try_into()?)
    }
}

#[action]
pub struct BidirectionalDigAction {
    pub outgoing: String,
    pub returning: String,
    pub new_area: String,
}

impl Action for BidirectionalDigAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!(
            "bidirectional-dig {:?} <-> {:?} '{:?}'",
            self.outgoing, self.returning, self.new_area
        );

        let (_, living, area) = surroundings.unpack();

        let new_area: Entity = build_entity()
            .area()
            .default_scope::<Occupyable>()?
            .default_scope::<Containing>()?
            .name(&self.new_area)
            .desc(&self.new_area)
            .try_into()?;
        let new_area = session.add_entity(new_area)?;

        tools::add_route(&area, &self.outgoing, &new_area)?;
        tools::add_route(&new_area, &self.returning, &area)?;

        match tools::navigate_between(&area, &new_area, &living)? {
            true => Ok(session.perform(Perform::Living {
                living,
                action: PerformAction::Instance(Rc::new(LookAction {})),
            })?),
            false => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, ToTaggedJson, DeserializeTagged)]
pub struct SaveQuickEditAction {
    pub key: EntityKey,
    pub copy: WorkingCopy,
}

impl SaveQuickEditAction {
    pub fn new(key: EntityKey, copy: WorkingCopy) -> Self {
        Self { key, copy }
    }

    pub fn new_template(key: EntityKey) -> Result<JsonTemplate, TaggedJsonError> {
        let copy = WorkingCopy::Markdown(JSON_TEMPLATE_VALUE_SENTINEL.to_owned());
        let template = Self { key, copy };

        Ok(template.to_tagged_json()?.into())
    }
}

impl Action for SaveQuickEditAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, _surroundings: &Surroundings) -> ReplyResult {
        info!("save:quick-edit {:?}", self.key);

        match session.entity(&LookupBy::Key(&self.key))? {
            Some(entity) => match &self.copy {
                WorkingCopy::Markdown(text) => {
                    let quick = QuickEdit::from_str(text)?;
                    let mut entity = entity.borrow_mut();
                    if let Some(name) = quick.name {
                        entity.set_name(&name)?;
                    }
                    if let Some(desc) = quick.desc {
                        entity.set_desc(&desc)?;
                    }

                    Ok(SimpleReply::Done.try_into()?)
                }
                _ => Err(anyhow::anyhow!("Save expected JSON working copy")),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, ToTaggedJson, DeserializeTagged)]
pub struct SaveEntityJsonAction {
    pub key: EntityKey,
    pub copy: WorkingCopy,
}

impl SaveEntityJsonAction {
    pub fn new(key: EntityKey, copy: WorkingCopy) -> Self {
        Self { key, copy }
    }

    pub fn new_template(key: EntityKey) -> Result<JsonTemplate, TaggedJsonError> {
        let copy = WorkingCopy::Json(JsonValue::String(JSON_TEMPLATE_VALUE_SENTINEL.to_owned()));
        let template = Self { key, copy };

        Ok(template.to_tagged_json()?.into())
    }
}

impl Action for SaveEntityJsonAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, _surroundings: &Surroundings) -> ReplyResult {
        info!("save:entity-json {:?}", self.key);

        match session.entity(&LookupBy::Key(&self.key))? {
            Some(entity) => match &self.copy {
                WorkingCopy::Json(value) => {
                    let replacing = Entity::from_value(value.clone())?;
                    entity.replace(replacing);

                    Ok(SimpleReply::Done.try_into()?)
                }
                _ => Err(anyhow::anyhow!("Save expected JSON working copy")),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct BuildAreaAction {
    pub name: String,
}

impl Action for BuildAreaAction {
    fn is_read_only() -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("build-area {:?}", self.name);

        let creator = surroundings.living();

        let new_area: Entity = build_entity()
            .area()
            .default_scope::<Containing>()?
            .default_scope::<Occupyable>()?
            .creator(creator.entity_ref())
            .name(&self.name)
            .try_into()?;

        let new_area = session.add_entity(new_area)?;

        remember(
            &creator,
            Utc::now(),
            MemoryEvent::Constructed(EntityEvent {
                key: new_area.key().clone(),
                gid: new_area.gid(),
                name: new_area.name()?.unwrap(),
            }),
        )?;

        info!("created {:?}", new_area);

        let observed = new_area_observation(creator, &new_area)?;
        let reply = Constructed::Area(observed);
        Ok(reply.try_into()?)
    }
}
