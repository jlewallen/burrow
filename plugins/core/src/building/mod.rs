use crate::library::plugin::*;

pub mod actions;
pub mod model;
pub mod parser;
#[cfg(test)]
mod tests;

#[derive(Default)]
pub struct BuildingPluginFactory {}

impl PluginFactory for BuildingPluginFactory {
    fn create_plugin(&self) -> Result<Box<dyn Plugin>> {
        Ok(Box::new(BuildingPlugin {}))
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

#[derive(Default)]
pub struct BuildingPlugin {}

impl Plugin for BuildingPlugin {
    fn plugin_key() -> &'static str
    where
        Self: Sized,
    {
        "building"
    }

    fn key(&self) -> &'static str {
        Self::plugin_key()
    }

    fn initialize(&mut self) -> Result<()> {
        Ok(())
    }

    fn middleware(&mut self) -> Result<Vec<Rc<dyn Middleware>>> {
        Ok(Vec::default())
    }

    fn deliver(&self, _incoming: &Incoming) -> Result<()> {
        Ok(())
    }

    fn stop(&self) -> Result<()> {
        Ok(())
    }
}

impl ParsesActions for BuildingPlugin {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        try_parsing(parser::EditActionParser {}, i)
            .or_else(|_| try_parsing(parser::DescribeActionParser {}, i))
            .or_else(|_| try_parsing(parser::DuplicateActionParser {}, i))
            .or_else(|_| try_parsing(parser::BidirectionalDigActionParser {}, i))
            .or_else(|_| try_parsing(parser::ObliterateActionParser {}, i))
            .or_else(|_| try_parsing(parser::MakeItemParser {}, i))
    }
}
