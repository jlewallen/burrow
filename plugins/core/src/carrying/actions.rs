use crate::{carrying::model::Carrying, library::actions::*, looking::model::Observe};

#[action]
pub struct HoldAction {
    pub item: Item,
}

impl Action for HoldAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("hold {:?}!", self.item);

        let (_, actor, area) = surroundings.unpack();

        match session.find_item(surroundings, &self.item)? {
            Some(holding) => match tools::move_between(&area, &actor, holding.clone())? {
                true => Ok(reply_ok(
                    actor.clone(),
                    Audience::Area(area.key().clone()),
                    Carrying::Held {
                        actor: (&actor).observe(&actor)?.expect("No observed entity"),
                        item: (&holding.entity()?)
                            .observe(&actor)?
                            .expect("No observed entity"),
                        area: (&area).observe(&actor)?.expect("No observed entity"),
                    },
                )?),
                false => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct DropAction {
    pub maybe_item: Option<Item>,
}

impl Action for DropAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("drop {:?}!", self.maybe_item);

        let (_, actor, area) = surroundings.unpack();

        match &self.maybe_item {
            Some(item) => match session.find_item(surroundings, item)? {
                Some(dropping) => match tools::move_between(&actor, &area, dropping.clone())? {
                    true => Ok(reply_ok(
                        actor.clone(),
                        Audience::Area(area.key().clone()),
                        Carrying::Dropped {
                            actor: (&actor).observe(&actor)?.expect("No observed entity"),
                            item: (&dropping.entity()?)
                                .observe(&actor)?
                                .expect("No observed entity"),
                            area: (&area).observe(&actor)?.expect("No observed entity"),
                        },
                    )?),
                    false => Ok(SimpleReply::NotFound.try_into()?),
                },
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct PutInsideAction {
    pub item: Item,
    pub vessel: Item,
}

impl Action for PutInsideAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("put-inside {:?} -> {:?}", self.item, self.vessel);

        let (_, _user, _area) = surroundings.unpack();

        match session.find_item(surroundings, &self.item)? {
            Some(item) => match session.find_item(surroundings, &self.vessel)? {
                Some(vessel) => {
                    let vessel = vessel.one()?;
                    if tools::is_container(&vessel)? {
                        let from = tools::container_of(&item.clone().one()?)?;
                        match tools::move_between(&from, &vessel, item)? {
                            true => Ok(SimpleReply::Done.try_into()?),
                            false => Ok(SimpleReply::NotFound.try_into()?),
                        }
                    } else {
                        Ok(SimpleReply::Impossible.try_into()?)
                    }
                }
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct TakeOutAction {
    pub item: Item,
    pub vessel: Item,
}

impl Action for TakeOutAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("take-out {:?} -> {:?}", self.item, self.vessel);

        let (_, user, _area) = surroundings.unpack();

        match session.find_item(surroundings, &self.vessel)? {
            Some(vessel) => {
                let vessel = vessel.one()?;
                if tools::is_container(&vessel)? {
                    match session.find_item(surroundings, &self.item)? {
                        Some(item) => match tools::move_between(&vessel, &user, item)? {
                            true => Ok(SimpleReply::Done.try_into()?),
                            false => Ok(SimpleReply::NotFound.try_into()?),
                        },
                        None => Ok(SimpleReply::NotFound.try_into()?),
                    }
                } else {
                    Ok(SimpleReply::Impossible.try_into()?)
                }
            }
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}

#[action]
pub struct GiveToAction {
    pub item: Item,
    pub receiver: Item,
}

impl Action for GiveToAction {
    fn is_read_only(&self) -> bool {
        false
    }

    fn perform(&self, session: SessionRef, surroundings: &Surroundings) -> ReplyResult {
        info!("give-to {:?} -> {:?}", self.item, self.receiver);

        let (_, user, _area) = surroundings.unpack();

        // I think there are very interesting permission related implications
        // here. For example, limiting third party access to your hands except
        // for key individuals.
        match session.find_item(surroundings, &self.item)? {
            Some(item) => match session.find_item(surroundings, &self.receiver)? {
                Some(receiver) => match tools::move_between(&user, &receiver.one()?, item)? {
                    true => Ok(SimpleReply::Done.try_into()?),
                    false => Ok(SimpleReply::NotFound.try_into()?),
                },
                None => Ok(SimpleReply::NotFound.try_into()?),
            },
            None => Ok(SimpleReply::NotFound.try_into()?),
        }
    }
}
