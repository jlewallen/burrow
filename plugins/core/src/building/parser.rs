use crate::library::parser::*;

use super::actions::{
    AddScopeAction, BidirectionalDigAction, BuildAreaAction, DuplicateAction, EditAction,
    EditRawAction, LimboAction, MakeItemAction, ObliterateAction,
};

pub struct EditActionParser {}

impl ParsesActions for EditActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = alt((
            map(
                preceded(pair(tag("edit raw"), spaces), noun_or_specific),
                |item| -> Box<dyn Action> { Box::new(EditRawAction { item }) },
            ),
            map(
                preceded(pair(tag("edit"), spaces), noun_or_specific),
                |item| -> Box<dyn Action> { Box::new(EditAction { item }) },
            ),
        ))(i)?;

        Ok(Some(action))
    }
}

pub struct MakeItemParser {}

impl ParsesActions for MakeItemParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(
                tuple((tag("@make"), spaces, tag("item"), spaces)),
                string_literal,
            ),
            |name| MakeItemAction { name: name.into() },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct DuplicateActionParser {}

impl ParsesActions for DuplicateActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@duplicate"), spaces), noun_or_specific),
            |item| DuplicateAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct LimboActionParser {}

impl ParsesActions for LimboActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(tag("@limbo"), |_| LimboAction {})(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct ObliterateActionParser {}

impl ParsesActions for ObliterateActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@obliterate"), spaces), noun_or_specific),
            |item| ObliterateAction { item },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct BidirectionalDigActionParser {}

impl ParsesActions for BidirectionalDigActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            tuple((
                preceded(pair(tag("@dig"), spaces), string_literal),
                preceded(pair(spaces, pair(tag("to"), spaces)), string_literal),
                preceded(pair(spaces, pair(tag("for"), spaces)), string_literal),
            )),
            |(outgoing, returning, new_area)| BidirectionalDigAction {
                outgoing: outgoing.into(),
                returning: returning.into(),
                new_area: new_area.into(),
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct ScopeActionParser {}

impl ParsesActions for ScopeActionParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(pair(tag("@scope"), spaces), camel_case_word),
            |scope_key| AddScopeAction {
                scope_key: scope_key.to_owned(),
            },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}

pub struct BuildAreaParser {}

impl ParsesActions for BuildAreaParser {
    fn try_parse_action(&self, i: &str) -> EvaluationResult {
        let (_, action) = map(
            preceded(tuple((tag("@build"), spaces)), string_literal),
            |name| BuildAreaAction { name: name.into() },
        )(i)?;

        Ok(Some(Box::new(action)))
    }
}
