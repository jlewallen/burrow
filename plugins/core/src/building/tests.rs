use super::parser::*;
use super::*;
use crate::building::actions::{SaveEntityJsonAction, SaveQuickEditAction};
use crate::building::model::QuickEdit;
use crate::fashion::model::Wearable;
use crate::library::tests::*;
use crate::{
    {carrying::model::Containing, looking::model::new_area_observation, tools},
    {BuildSurroundings, QuickThing},
};

#[test]
fn it_fails_to_edit_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_fails_to_edit_raw_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit raw rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_edits_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit broom")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    Ok(())
}

#[test]
fn it_edits_raw_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit raw broom")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    Ok(())
}

#[test]
fn it_fails_to_edit_items_by_missing_gid() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit #1201")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_edits_items_by_gid() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(EditActionParser {}, "edit #2")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    insta::assert_json_snapshot!(reply.to_debug_json()?);

    Ok(())
}

#[test]
fn it_fails_to_duplicate_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .ground(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(DuplicateActionParser {}, "@duplicate rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_duplicates_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(DuplicateActionParser {}, "@duplicate broom")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);
    assert_eq!(person.scope::<Containing>()?.holding.len(), 1);
    assert_eq!(
        tools::quantity(&person.scope::<Containing>()?.holding[0].to_entry()?)?,
        2.0
    );

    Ok(())
}

#[test]
fn it_fails_to_obliterate_unknown_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(ObliterateActionParser {}, "@obliterate rake")?;
    let action = action.unwrap();
    let reply = action.perform(session, &surroundings)?;

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::NotFound);

    Ok(())
}

#[test]
fn it_obliterates_items_named() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build
        .hands(vec![QuickThing::Object("Cool Broom")])
        .build()?;

    let action = try_parsing(ObliterateActionParser {}, "@obliterate broom")?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_world, person, area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);
    // It's not enough just to check this, but why not given how easy.
    // Should actually verify it's deleted.
    assert_eq!(person.scope::<Containing>()?.holding.len(), 0);
    assert_eq!(area.scope::<Containing>()?.holding.len(), 0);

    build.flush()?;

    Ok(())
}

#[test]
fn it_digs_bidirectionally() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;

    let action = try_parsing(
        BidirectionalDigActionParser {},
        r#"@dig "North Exit" to "South Exit" for "New Area""#,
    )?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, _area) = surroundings.unpack();

    // Not the best way of finding the constructed area.
    let destination = session
        .entry(&LookupBy::Gid(&EntityGid::new(4)))?
        .ok_or(DomainError::EntityNotFound)?;

    let reply: AreaObservation = reply.json_as()?;
    assert_eq!(reply, new_area_observation(&living, &destination)?);

    Ok(())
}

#[test]
fn it_makes_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;

    let action = try_parsing(MakeItemParser {}, r#"@make item "Blue Rake""#)?;
    let action = action.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(living.scope::<Containing>()?.holding.len(), 1);

    Ok(())
}

#[test]
fn it_saves_changes_to_description() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;
    let (_world, _living, _area) = surroundings.unpack();

    let description = "Would be really weird if this was the original description".to_owned();
    let mut quick_edit = QuickEdit::default();
    quick_edit.name = Some("NAME!".to_owned());
    quick_edit.desc = Some(description.to_owned());

    let action = Box::new(SaveQuickEditAction {
        key: EntityKey::new("world"),
        copy: WorkingCopy::Markdown(quick_edit.to_string()),
    });
    let reply = action.perform(session.clone(), &surroundings)?;
    let (world, _living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(world.desc()?.unwrap(), description.as_str());

    Ok(())
}

#[test]
fn it_saves_changes_to_whole_entities() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let (session, surroundings) = build.plain().build()?;
    let (_, living, area) = surroundings.unpack();

    let original = living.entity();
    let original = original.to_json_value()?;

    let action = Box::new(SaveEntityJsonAction {
        key: area.key().clone(),
        copy: WorkingCopy::Json(original),
    });
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, _living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    // TODO Would be really nice to have some assurances here, even though
    // I'm wondering how often this will actually get used.

    Ok(())
}

#[test]
fn it_adds_scopes_to_solo_held_items() -> Result<()> {
    let mut build = BuildSurroundings::new()?;
    let jacket = build.entity()?.named("Jacket")?.carryable()?.into_entry()?;

    assert!(!jacket.has_scope::<Wearable>()?);

    let (session, surroundings) = build
        .plain()
        .hands(vec![QuickThing::Actual(jacket.clone())])
        .build()?;

    let action = try_parsing(ScopeActionParser {}, r#"@scope wearable"#)?.unwrap();
    let reply = action.perform(session.clone(), &surroundings)?;
    let (_, living, _area) = surroundings.unpack();

    let reply: SimpleReply = reply.json_as()?;
    assert_eq!(reply, SimpleReply::Done);

    assert_eq!(living.scope::<Containing>()?.holding.len(), 1);

    assert!(jacket.has_scope::<Wearable>()?);

    Ok(())
}
