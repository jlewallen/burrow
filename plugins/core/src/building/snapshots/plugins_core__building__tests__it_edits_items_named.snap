---
source: plugins/core/src/building/tests.rs
expression: reply.to_debug_json()?
---
{
  "reply": {
    "taggedJson": [
      "editorReply",
      {
        "editing": {
          "markdown": "Cool Broom\n\n~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n\n"
        },
        "key": "E-3",
        "save": {
          "saveQuickEditAction": {
            "copy": {
              "markdown": "!#$value"
            },
            "key": "E-3"
          }
        }
      }
    ]
  }
}
